src/sender.ts

```diff
@@ -440,64 +433,17 @@ class Sender {
-      // This case should ideally not happen if logic is correct, means we tried to compact more than available
-      this.position = 0;
-    }
-    // If bytesSent is 0 or negative, or if no actual data was at the start of the buffer to be shifted,
-    // this.position effectively remains the same relative to the start of new data.
-
-    this.endOfLastRow = Math.max(0, this.endOfLastRow - bytesSent);
-    // Ensure endOfLastRow is also shifted if it was within the compacted area, 
-    // or reset if it pointed to data that's now gone. 
-    // If new rows were added while flushing, endOfLastRow would be > position post-compaction of old data.
-    // This needs careful handling if new data is added *during* an async flush.
-    // For now, we assume endOfLastRow is relative to the data just flushed.
-    // A simpler approach might be to always set this.endOfLastRow = 0 after a successful flush, 
-    // as startNewRow() will set it correctly for the *next* new row.
-    // However, if a flush doesn't clear all pending complete rows, this needs to be accurate.
-    // The current `flush` logic sends up to `this.endOfLastRow`, so after sending `dataAmountToSend`
-    // (which was `this.endOfLastRow` at the time of prepping the flush), the new `this.endOfLastRow`
-    // should effectively be 0 relative to the start of the compacted buffer, until a new row is started.
-
-    this.lastFlushTime = Date.now();
-    this.pendingRowCount = 0; // Reset after successful flush
-    // If autoFlush was triggered by row count, this reset is crucial.
-    // If triggered by interval, this is also fine.
-  }
-
-  /**
-   * @ignore
-   * Executes the actual data sending logic (HTTP or TCP).
-   * This is called by the `flush` method, wrapped in the promise chain.
-   * @return {Promise<boolean>} Resolves to true if data was sent.
+   * Sends the buffer's content to the database and compacts the buffer.
+   * If the last row is not finished it stays in the sender's buffer.
+   *
+   * @return {Promise<boolean>} Resolves to true when there was data in the buffer to send, and it was sent successfully.
    */
-  private async _executeFlush(): Promise<boolean> {
-    const dataAmountToSend = this.endOfLastRow;
-    if (dataAmountToSend <= 0) {
+  async flush(): Promise<boolean> {
+    const dataToSend = this.toBufferNew(this.endOfLastRow);
+    if (!dataToSend) {
       return false; // Nothing to send
     }
 
-    // Use toBufferView to get a reference, actual data copy for sending happens based on protocol needs
-    const dataView = this.toBufferView(dataAmountToSend);
-    if (!dataView) {
-      return false; // Should not happen if dataAmountToSend > 0, but a safe check
-    }
-
-    // Create a copy for sending to avoid issues if the underlying buffer changes
-    // This is especially important for async operations.
-    const dataToSend = Buffer.allocUnsafe(dataView.length);
-    dataView.copy(dataToSend);
-
```

Comment:
The below code seems to create a `RetryAgent` per each `flush()` call. Can't we reuse single agent across all calls?
```js
const dispatcher = new RetryAgent(this.agent, {
```

Reply:
Makes sense.

Reply:
addressed in https://github.com/questdb/nodejs-questdb-client/pull/48